tensorflow/python/keras/_impl/keras/callbacks.py

```diff
@@ -712,7 +712,7 @@ def set_model(self, model):
         for weight in layer.weights:
           mapped_weight_name = weight.name.replace(':', '_')
           tf_summary.histogram(mapped_weight_name, weight)
-          if self.write_grads:
+          if self.write_grads and weight in layer.trainable_weights:
```

Comment:
Prefer creating a set beforehand to avoid repeated list lookups:

```
trainable_weights = set(layer.trainable_weights)
for weight in layer.weights:
    ...
    if self.write_grads and weight in trainable_weights:
```

Another option is to factor out the `self.write_grads` loop:

```python
if self.write_grads:
    for weight in model.trainable_weights:
       ...
```

Reply:
I prefer to factor out the code, and could you take a look again?